java/dataset/src/main/cpp/jni_wrapper.cc

```diff
@@ -83,6 +83,40 @@ void ThrowIfError(const arrow::Status& status) {
   }
 }
 
+class JNIEnvGuard {
+ public:
+  explicit JNIEnvGuard(JavaVM* vm) : vm_(vm), should_detach_(false) {
```

Comment:
This should initialize in the order the fields are declared, and env_ should be initialized. Probably better to re-order the fields such that the bool is the last field though to minimize padding.

Reply:
env_ is initialized at the end of constructor...

Reply:
The compiler emits a build warning if it isn't done in the initializer list (even though you are setting it later).

Reply:
Ok fixed, I didn't get (or I missed) the c++ compiler's warnings with archery ...